instance/tests/models/test_openedx_monitoring_mixins.py

```diff
@@ -93,11 +104,86 @@ def test_enable_monitoring(self, additional_monitoring_emails, expected_monitor_
+
+        for add_call in mock_newrelic.add_alert_condition.call_args_list:
+            created_monitor_ids.add(add_call[0][1])
+
+        for add_call in mock_newrelic.add_email_notification_channel.call_args_list:
+            list_of_emails_added.append(add_call[0][0])
+
         self.assertEqual(set(monitor_ids), created_monitor_ids)
+        self.assertEqual(set(list_of_emails_added), set(expected_monitor_emails))
+        mock_newrelic.add_notification_channels_to_policy.assert_called_with(
+            1, list(range(len(expected_monitor_emails)))
+        )
+
+    @patch('instance.models.mixins.openedx_monitoring.newrelic')
+    def test_enable_monitoring_for_pre_new_relic_alerts_instances(self, mock_newrelic):
+        """
+        Check that the monitoring is enabled properly for the instances with monitoring
+        enabled using the pre-New Relic alerts code.
+        """
+        instance = OpenEdXInstanceFactory()
+        monitor_ids = [str(uuid4()) for i in range(len(instance._urls_to_monitor))]
+        for index, monitored_url in enumerate(instance._urls_to_monitor.keys()):
+            NewRelicAvailabilityMonitor.objects.create(instance=instance, pk=monitor_ids[index])
+
+        existing_monitors = NewRelicAvailabilityMonitor.objects.filter(instance=instance)
+
+        self.assertEqual(existing_monitors.count(), len(instance._urls_to_monitor))
+        self.assertEqual(NewRelicAlertPolicy.objects.filter(instance=instance).count(), 0)
+        mock_newrelic.get_synthetics_monitor.side_effect = [{'uri': url} for url in instance._urls_to_monitor]
+        mock_newrelic.get_synthetics_notification_emails.return_value = ['admin@opencraft.com']
+        mock_newrelic.add_alert_policy.return_value = 1
+        mock_newrelic.add_alert_condition.side_effect = list(range(4))
+        mock_newrelic.add_email_notification_channel.return_value = 1
+        instance.enable_monitoring()
+        self.assertEqual(set(NewRelicAvailabilityMonitor.objects.filter(instance=instance)), set(existing_monitors))
+        self.assertEqual(NewRelicAlertPolicy.objects.filter(instance=instance).count(), 1)
+        self.assertEqual(
+            NewRelicEmailNotificationChannel.objects.get(new_relic_alert_policies__id=1).email,
+            'admin@opencraft.com'
+        )
+        self.assertEqual(
+            NewRelicAlertCondition.objects.filter(alert_policy__id=1, monitor__id__in=monitor_ids).count(), 4
+        )
+
+    @patch('instance.models.mixins.openedx_monitoring.newrelic')
+    def test_enable_monitoring_does_not_skip_any_new_relic_alerts_code_on_retry_after_error(self, mock_newrelic):
+        """
+        Check that the enable_monitoring doesn't skip the creation of monitoring resources when a previous attempt
+        to enable monitoring failed with an error.
+        """
+        instance = OpenEdXInstanceFactory()
+
+        class CustomException(Exception):
+            pass
+
+        def check(instance):
```

Comment:
@giovannicimolin this function isn't very clear to me. It seems that the first invocation will also throw `CustomException` (it seems that all invocations will do), so why are we using a broad „catch Exception“ as if we didn't know whether it will succeed or not? And if we expect 1+10 failures, why not directly loop 11 times? (or 10, to round it).
I also added a doctest [in my branch](https://github.com/open-craft/opencraft/pull/436/commits/892c199fba57e5a918de6b2e52218a881a5ca98d) (still untested). I'll write some questions in the JIRA task.

Reply:
@clemente, the idea behind this test is - if some step in enabling the monitoring, particularly in the new changes added in this PR, for an instance fails, retrying the same should not skip the failed step. The code uses only the data in the models for checking if a New Relic resource (alert policy, alert condition, notification channel) exists or not. While testing the code changes in this PR, I found that I had introduced a bug where the model instance was getting created before the actual New Relic resource was created and when the New Relic API call failed due to some reason, the retry skipped the code to try creating it again because the model instance exists.

Since that code was confusing, I added this test to ensure that if a New Relic API call fails due to some reason, a retry will still traverse through the same code path on every retry till it succeeds.